src/esp/gfx/FisheyeShader.h

```diff
@@ -0,0 +1,118 @@
+#include <Corrade/Utility/Macros.h>
+#include <Magnum/GL/AbstractShaderProgram.h>
+#include <Magnum/GL/CubeMapTexture.h>
+#include <Magnum/Shaders/Generic.h>
+
+#include "esp/core/esp.h"
+
+namespace esp {
+namespace gfx {
+namespace fisheyeShaderTexUnitSpace {
+enum TextureUnit : uint8_t {
+  Color = 0,
+  Depth = 1,
+  // TODO
+  // ObjectId = 2,
+};
+}
+// Interface class for various fisheye camera shaders, such as "Double Sphere
+// Camera", "Field-of-View Camera", etc.
+class FisheyeShader : public Magnum::GL::AbstractShaderProgram {
+ public:
+  enum : Magnum::UnsignedInt {
+    /**
+     * Color shader output. @ref shaders-generic "Generic output",
+     * present always. Expects three- or four-component floating-point
+     * or normalized buffer attachment.
+     */
+    ColorOutput = Magnum::Shaders::Generic3D::ColorOutput,
+
+    // TODO
+    /**
+     * Object ID shader output. @ref shaders-generic "Generic output",
+     * present only if @ref Flag::ObjectId is set. Expects a
+     * single-component unsigned integral attachment. Writes the value
+     * set in @ref setObjectId() there.
+     */
+    // ObjectIdOutput = Magnum::Shaders::Generic3D::ObjectIdOutput,
+  };
+
+  /**
+   * @brief Flag
+   *
+   * @see @ref Flags, @ref flags()
+   */
+  enum class Flag : Magnum::UnsignedShort {
+    /**
+     * cubemap color texture
+     */
+    ColorTexture = 1 << 0,
+    /**
+     * cubemap depth texture
+     */
+    DepthTexture = 1 << 1,
+    // ObjectIdTexture = 1 << 2,
+  };
+
+  typedef Corrade::Containers::EnumSet<Flag> Flags;
+
+  /**
+   * @brief constructor
+   * @param[in] flags fisheye shader flags
+   */
+  explicit FisheyeShader(Flags flags = {Flag::ColorTexture});
+
+  ~FisheyeShader() override = default;
+
+  /** @brief Flags */
+  Flags flags() const { return flags_; }
+
+  /**
+   * @brief bind cubemap color texture
+   * @param[in] texture cubemap color texture
+   */
+  virtual FisheyeShader& bindColorTexture(Magnum::GL::CubeMapTexture& texture);
+  /**
+   * @brief bind cubemap depth texture
+   * @param[in] texture cubemap depth texture
+   */
+  virtual FisheyeShader& bindDepthTexture(Magnum::GL::CubeMapTexture& texture);
+  // virtual FisheyeShader& bindObjectIdTexture(Magnum::GL::Texture2D&
+  // texture);
+
+ protected:
+  Flags flags_;
+
+  // ======= uniforms =======
+  // it hurts the performance to call glGetUniformLocation() every frame due
+  // to string operations. therefore, cache the locations in the constructor
+  // using the function @ref cacheUniforms()
+
+  // common material uniforms
+
+  /**
+   * @brief cache the uniform locations
+   * NOTE: Subclass must implement this function, and call it at the end of
+   * the constructor
+   */
+  virtual void cacheUniforms() = 0;
+
+  /**
+   * @brief set texture binding points in the shader
+   * NOTE: Subclass must implement this function, and call it at the end of
+   * the constructor
+   */
+  virtual void setTextureBindingPoints() = 0;
```

Comment:
I don't understand why these need to exist at all... the base class has no way of invoking these from its own constructor anyway. To me it feels like `DoubleSphereCameraShader` should just put the contents of these functions inline into the constructor, like all other shaders do.

Reply:
My primary concern was with the `virtual`ity, there's no point in having these virtual, as the base constructor can't call them anyway, and it just makes the class heavier for no reason. Yes, making them pure virtual is _a_ way to enforce the subclass author to provide implementations for those, but it ends there, and the author still has to look around and check *what* the functions need to do and *where* these need to be called, and at that point it could as well be just a comment saying 

> Implement a subclass the same way as `DoubleSphereFisheyeShader` please or @bigbike will get upset.

Apart from the `virtual` aspect -- I know your position on this and I remember we were discussing this quite extensively in the past. However, with shader construction, the steps have to be done in a very rigid manner and I feel this just increases potential for bugs instead of reducing it, as it's easier to:

- forget calling them at the end of the constructor after a larger refactor
- implement them wrong just because they live in a different portion of the file and so the mistakes that you'd immediately see when the code was directly inlined are now non-obvious
- call them before the program was linked -- "she can do it without needing to know any details" is simply not true, the developer *has to know* what these functions exactly do to call them at the right place
- call them outside of the constructor again, repeatedly, in each setter, for no reason at all, just because it seemed to fix some bug

but on the other hand it's unnecessarily hard to transfer any state from the constructor to these (for example, *what if* the constructor has some complex calculated state which affects what textures are fetched, or the uniform names are dynamic etc? now the class has to save that to a private memeber that's not used otherwise...).

Looking at it from a completely different angle, [init functions can be considered as an antipattern](https://softwareengineering.stackexchange.com/q/334970).

- - -

I probably shared this link already, this is John Carmack's view on the topic and he sums up the pros and cons of both approaches better than I could ever do: http://number-none.com/blow/john_carmack_on_inlined_code.html